### DISCUSSION TITLE
Next.js module was multi-trigged in getStaticProps

### DISCUSSION THREAD

**Discussion by PrinOrange** [vercel/next.js#71944] [Help] [answered]
### Summary

In Next.js, if a module has `export const obj = func()` , does the `func` function need to **be triggered every time** other modules import this `obj`? Then why does func need to be triggered every time when it is referenced in `getStaticProps` on different pages?

For example, we have a module

lib/my-module.ts
```ts
const func = async () => {
      console.log("trigged");
      const data = await fetchData();
      return data;
}
export const obj = await func();
```

Then we refer the `obj` in every page's `getStaticProps`,

In home page
```tsx
import {obj} from './lib/my-module.ts'
export default const HomePage = (props) => {
      return <main/>
}
const getStaticProps = async () => {
      const data = obj
      return {props: {data:obj} }
}
```
And in another page
```tsx
import {obj} from './lib/my-module.ts'
export default const AnotherPage = (props) => {
      return <main/>
}
const getStaticProps = async () => {
      const data = obj
      return {props: {data:obj} }
}
```

We noticed that during the building the function `func` was trigged twice.
```
trigged
trigged
```

However, in node.js, aren't modules only loaded once and put into memory and never triggered again? Why is the module loaded multiple times when next.js references it in `getStaticProps` on different pages?

If `func` is a time-consuming function, this will cause multiple unnecessary loads, seriously slowing down the build time.
Is there any way to make func only trigger once?

Thanks

**Comment by icyJoseph** [comment, marked as answer]
I think this is because workers are used to build the pages. So each worker, creates its own instance so to speak. Try setting `experimental.workerThreads` set to false, or `experimental.cpus` set to 1.

I'll try to reproduce the issue, and see if that fix works.

**Reply by icyJoseph** [reply]
So, I managed to reproduce, and got `console.count('func')` to print up to three times.

Setting `cpus` to 1, seems to cut it down to 2 times only. Will keep on testing and updating this comment.

## Edit

Printing PID of the root process, and the process loading the module:

```js
const func = () => {
  console.count("func");
  console.log("Current:", process.pid, "Parent:", process.ppid);

  return {};
};

export const obj = func();

```

We get:

```shell
Root Parent 38170
## other logs
 ✓ Collecting page data
func: 1
Current: 38173 Parent: 38170
 ✓ Collecting page data
func: 1
Current: 38174 Parent: 38170
```

Using `cpus` set to 1, if we just let it pick `cpus`:

```shell
Root Parent 38918
## other logs
 ✓ Collecting page data
func: 1
Current: 38963 Parent: 38918
func: 1
Current: 38964 Parent: 38918
 ✓ Collecting page data
func: 1
Current: 38968 Parent: 38918
```

You can see, that there's three child processes, this time around, whereas before we had one.

I don't remember which config to use, to get everything to be done in one process though...

**Reply by PrinOrange** [reply]
It works temporally, make do with it for now. Thank you.

These are the relevant issues that have same problem with mine. https://github.com/vercel/next.js/issues/65350 and https://github.com/vercel/next.js/issues/55885
I also do some research but found nothing better solution. The solutions are not working as far, still a tricky problem.

My purpose is set a singleton in a module, and every page's `getStaticProps` should use same one singleton during building time. It means that the module singleton is supposed to be initialized once. Because the singleton initializing is very time-costing. 

Perhaps, `getStaticProps` of next.js is supposed to provide an entrance to pass singleton to each worker for use during build.

**Reply by icyJoseph** [reply]
Yeah, I have done a somewhat similar pattern, although pure, many years ago, and it worked fine I think. I feel like something must have changed 🤔
